buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -229,18 +229,111 @@ public static int hashCode(ByteBuf buffer) {
      * Returns the reader index of needle in haystack, or -1 if needle is not in haystack.
      */
     public static int indexOf(ByteBuf needle, ByteBuf haystack) {
-        // TODO: maybe use Boyer Moore for efficiency.
-        int attempts = haystack.readableBytes() - needle.readableBytes() + 1;
-        for (int i = 0; i < attempts; i++) {
-            if (equals(needle, needle.readerIndex(),
-                       haystack, haystack.readerIndex() + i,
-                       needle.readableBytes())) {
-                return haystack.readerIndex() + i;
+        if (haystack == null || needle == null) {
+            return -1;
+        }
+
+        if (needle.readableBytes() > haystack.readableBytes()) {
+            return -1;
+        }
+
+        int n = haystack.readableBytes();
+        int m = needle.readableBytes();
+        if (m == 0) {
+            return 0;
+        }
+
+        int i = maxSuf(needle, m, true);
+        int j = maxSuf(needle, m, false);
+        int ell = Math.max(i, j);
```

Comment:
Please replace the `i` and `j` variables in the `max` call with the `maxSuf` calls.

As it is right now, the `i` and `j` variables are used for two different purposes, which is confusing to read. We can instead declare them closer to where they are really used. That would make the code look more like Java and less like C.